src/project/manifest.rs

```diff
@@ -14,8 +17,554 @@ use serde_with::{serde_as, DeserializeAs, DisplayFromStr, PickFirst};
+/// The manifest data is represented as a [`ProjectManifest`] struct for easy manipulation.
+/// Owned by the [`Project`] struct, which governs its usage.
+///
+#[derive(Debug, Clone)]
+pub struct Manifest {
+    /// The path to the manifest file
+    pub path: PathBuf,
+
+    /// The raw contents of the manifest file
+    pub contents: String,
+
+    /// Editable toml document
+    pub document: toml_edit::Document,
+
+    /// The parsed manifest
+    pub parsed: ProjectManifest,
+}
+
+impl Manifest {
+    /// Create a new manifest from a path
+    pub fn from_path(path: PathBuf) -> miette::Result<Self> {
+        let contents = std::fs::read_to_string(&path).into_diagnostic()?;
+        Self::from_str(
+            path.parent().expect("Path should always have a parent"),
+            contents,
+        )
+    }
+
+    /// Create a new manifest from a string
+    pub fn from_str(root: &Path, contents: impl Into<String>) -> miette::Result<Self> {
+        let contents = contents.into();
+        let (manifest, document) = match toml_edit::de::from_str::<ProjectManifest>(&contents)
+            .map_err(TomlError::from)
+            .and_then(|manifest| contents.parse::<Document>().map(|doc| (manifest, doc)))
+        {
+            Ok(result) => result,
+            Err(e) => {
+                if let Some(span) = e.span() {
+                    return Err(miette::miette!(
+                        labels = vec![LabeledSpan::at(span, e.message())],
+                        "failed to parse project manifest"
+                    )
+                    .with_source_code(NamedSource::new(consts::PROJECT_MANIFEST, contents)));
+                } else {
+                    return Err(e).into_diagnostic();
+                }
+            }
+        };
+
+        // Validate the contents of the manifest
+        manifest.validate(
+            NamedSource::new(consts::PROJECT_MANIFEST, contents.to_owned()),
+            root,
+        )?;
+
+        // Notify the user that pypi-dependencies are still experimental
+        if manifest
+            .pypi_dependencies
+            .as_ref()
+            .map_or(false, |deps| !deps.is_empty())
+        {
+            tracing::warn!("ALPHA feature enabled!\n\nIt looks like your project contains `[pypi-dependencies]`. This feature is currently still in an ALPHA state!\n\nYou may encounter bugs or weird behavior. Please report any and all issues you encounter on our github repository:\n\n\thttps://github.com/prefix-dev/pixi.\n");
+        }
+
+        Ok(Self {
+            path: root.join(consts::PROJECT_MANIFEST),
+            contents,
+            document,
+            parsed: manifest,
+        })
+    }
+
+    /// Save the manifest to the file and update the contents
+    pub fn save(&mut self) -> miette::Result<()> {
+        self.contents = self.document.to_string();
+        std::fs::write(&self.path, self.contents.clone()).into_diagnostic()?;
+        Ok(())
+    }
```

Comment:
Can't see from github PR but does `&self` need to be mutable?

Reply:
I'm overwriting the `self.contents` that is what i needed the mutability for.